### PR TITLE
Make /boot/initrd.img-$KVER path configurable (default still /boot/initramfs-$KVER.img)

### LINKED ISSUE #99: detect and generate initrd.img-${kernel}
https://www.freedesktop.org/wiki/Specifications/BootLoaderSpec/ refers to initramfs as "initrd".

One of the most popular distro's (Debian) expects /boot/initrd.img-${kernel} instead of initramfs-${kernel}.img, see https://kernel-team.pages.debian.net/kernel-handbook/ch-packaging.html#s-arch-dep

Let's look for a way to detect in a reliable way what should be the default output file that is more inclusive.

Possibly related openSUSE patch downstream: https://github.com/dracutdevs/dracut/commit/f68649cf55ca4545ed8304435c9c7315846c45a4

### PR DESCRIPTION
**Pull request by nabijaczleweli**
To get the correct setup, run `./configure --initrd-prefix initrd.img --initrd-suffix ""`

## Checklist
- [x] I have tested it locally
- [x] I have reviewed and updated any documentation if relevant
- [x] I am providing new code and test(s) for it

Fixes #99

### CODE REVIEW

File: dracut-initramfs-restore.sh.in [hunk 1]
```diff
@@ -39,8 +39,8 @@ elif [[ -d /boot/efi/loader/entries || -L /boot/efi/loader/entries ]] \
     IMG="/boot/efi/$MACHINE_ID/$KERNEL_VERSION/initrd"
 elif [[ -f /lib/modules/${KERNEL_VERSION}/initrd ]]; then
     IMG="/lib/modules/${KERNEL_VERSION}/initrd"
-elif [[ -f /boot/initramfs-${KERNEL_VERSION}.img ]]; then
-    IMG="/boot/initramfs-${KERNEL_VERSION}.img"
+elif [[ -f /boot/@INITRDPREFIX@-${KERNEL_VERSION}@INITRDSUFFIX@ ]]; then
+    IMG="/boot/@INITRDPREFIX@-${KERNEL_VERSION}@INITRDSUFFIX@"
```
Comment (bdrung) [inline dracut-initramfs-restore.sh.in:43]: For full flexibility: What do you think about moving `-` also to the prefix (i.e. `initrdprefix=initramfs-`)?

File: dracut.html [hunk 2]
```diff
@@ -4149,4 +4149,4 @@
                                           switch-root</pre></div><div class="section"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a id="_author_4"></a>AUTHOR</h2></div></div></div><p>Harald Hoyer</p></div><div class="section"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a id="_see_also_6"></a>SEE ALSO</h2></div></div></div><p><span class="strong"><strong>dracut</strong></span>(8) <span class="strong"><strong>bootup</strong></span>(7)</p></div></div><div class="appendix"><div class="titlepage"><div><div><h2 class="title"><a id="_license"></a>Appendix A. License</h2></div></div></div><p>This work is licensed under the Creative Commons Attribution/Share-Alike
 License. To view a copy of this license, visit
 <a class="ulink" href="http://creativecommons.org/licenses/by-sa/3.0/" target="_top">http://creativecommons.org/licenses/by-sa/3.0/</a> or send a letter to Creative
-Commons, 559 Nathan Abbott Way, Stanford, California 94305, USA.</p></div></div></div></body></html>
+Commons, 559 Nathan Abbott Way, Stanford, California 94305, USA.</p></div></div></div></body></html>
```
Comment (bdrung) [inline dracut.html:4152]: Trailing newline is stripped?

**Review by bdrung — CHANGES_REQUESTED** [review]
Looks good (except for a small remark).

File: dracut.html (hunk 2 above)
Comment (nabijaczleweli) [reply dracut.html:4152]: that's what my asciidoc regenerated (why a non-stable generated file is checked in is beyond me but whatever). dropped

File: dracut-initramfs-restore.sh.in (hunk 1 above)
Comment (nabijaczleweli) [reply dracut-initramfs-restore.sh.in]: sure. applied

**Review by bdrung — APPROVED** [review]
Looks good from my side now.

**Review by freswa — APPROVED** [review]
lgtm

**Comment by pvalena** [comment]
Do we really want to make it so, that you have to build the project to test / consume it? Up until this point you could just unpack (clone) it and run it straight away....

**Comment by pvalena** [comment]
In a sense, yes, because if you have any of those already installed (and the code doesn't change that much), you just drop the shell scripts (or run them from new location).... anyway, as we should really test the "build" (e.g. rpm, shift left), I don't mind. Just pointing out, that I've heard about such use-cases.

**Comment by bdrung** [comment]
> Just pointing out, as I've heard about such use-case.

Debian carries a patch to change the initramfs file name. Being able to drop that patch is a benefit.

**Comment by Mrfai** [comment]
I don't think that the documentation needs to be totally adjusted to
every single path for each distribution. All people should know that
pathes may be slightly different on their distribution.


>>>>> On Tue, 09 Apr 2024 19:55:51 -0700, Laszlo Gombos ***@***.***> said:

    > I do not yet know how, but I hope to rephrase the manual in a way that we do not need to change the manual for each distribution (and
    > still get rid of the patch for Debian based distributions). In principle, we want the same documentation for each distribution.

    > I think once we solve the documentation issue, the code changes themselves will not be problem.

    > Wouldn't an entry in dracut.conf be less invasive ? The documentation could than say something along the lines of

    >     /boot/initramfs-.img. or the distribution default configured.

**Comment by nabijaczleweli** [comment]
This is crazy-talk. Just make the manual have the correct information inside, it's not hard. asciidoc has explicit support for this, and we already use it for the version.

File: dracut-initramfs-restore.sh.in (hunk 1 above)
Comment (Conan-Kudo) [reply dracut-initramfs-restore.sh.in]: I would actually prefer _not_ doing that. Is there a compelling reason to make the separators inconsistent?

**Comment by Conan-Kudo** [comment]
I would also like to challenge the idea that we should have different names for the initramfs files. This has been a major pain for downstream consumers of distributions that try to build automation. For example @schaefi and I have gone through a lot of pain trying to figure out how to discover and "use" the correct file path.

Is there a particularly compelling reason that Debian, Arch, Gentoo, etc. should be using different names for this?

File: dracut-initramfs-restore.sh.in (hunk 1 above)
Comment (nabijaczleweli) [reply dracut-initramfs-restore.sh.in]: What does "inconsistent" mean here. If you want to have dracut to make `6.1.0-12-amd64.initrd` then that's valid and not possible with a partially-fixed prefix.

**Comment by nabijaczleweli** [comment]
What's the premise of your post? The names are already different because they include the version. Don't see how this changes anything. Unrelated programs being RHELpilled isn't salient here.

"Is there a reason for supporting all systems as part of upstream configuration instead of requiring downstream patches?" Yes. It's so you support it upstream instead of requiring downstream patches from everyone except fedora.

**Comment by Conan-Kudo** [comment]
Please refrain from being antagonistic here. As a matter of fact, I'm actually one of the downstream dracut maintainers for openSUSE and work on this stuff in Mageia as well. I'm not very involved at this level in Fedora. Nor do I work for Red Hat either.

I am coming from the perspective of someone that develops tools that work in a cross-distribution way. This is one of those differences that causes a lot of pain for people trying to build upon different distributions and it has yet to be made clear why we should have those differences in the first place.

**Comment by bdrung** [comment]
Speaking with my Ubuntu hat on, unifying on one name for the initramfs file would be an option as well. I agree that having different names is not a good idea in itself.

**Comment by Conan-Kudo** [comment]
With my upstream hat in multiple projects (kiwi, livecd-tools, etc.) that consume initramfs stuff, I would rather prefer we agreed to use the path that dracut has used out of the gate (`initramfs-$KVER.img`). It makes life considerably easier for consumers of initramfs tooling to do the right thing.

Offhand, I know that the following distributions customize the name via patching:

* Debian/Ubuntu (`initrd.img-$KVER`)
* Mageia (`initrd-$KVER.img`)
* openSUSE (`initrd-$KVER`)

No one else I can find changes the name. 

At least from my side in Mageia, I would prefer to drop the patch and use upstream naming, it's just a matter of identifying if there's anything else left to fix to let us do it. When @danimo and I were more actively working through the openSUSE dracut package, we also intended to drop our patch on the initramfs file name, and I'd like to see if we can still do that. That just leaves the Debian family, which I would like to see if they could be convinced to use the `initramfs-$KVER.img` form too.

(Not that it should matter that much, but when Fedora switched from mkinitrd to dracut, we also gave up our custom name for the initramfs image, which was similar to Mageia's.)

**Comment by nabijaczleweli** [comment]
It's wild to see that when receiving a patch that all downstreams carry, your response is "actually the downstreams are wrong and my default is objectively correct (it matches fedora policy). also change your 20-year policy to my one because i dont like yours". The former used to be understandable under redhat management at least. Utterly delusional position backed up by "oh i had to look at three(!) policies to guess where the initrd for the current kernel was when porting my program outside of fedora (my new program is more important than all users' existing programs and policy). if i force downstreams to keep carrying this patch this will stop being a porting issue for me".

If anyone is using a "custom name" here (again, insane way to frame "not the fedora default"), it's dracut – initramfs-tools produced `/boot/initrd.img-$kver` in 2005-09-29 (which probably means more users *and* since earlier, for obvious reasons (it's initrd.img for the given kernel version), but this one's the first one on snapshot.d.o).

If Laszlo wants to rebase and re-post this as a new PR then he's welcome to. I no longer care to see the peanut gallery.

**Comment by zdykstra** [comment]
It's unfortunate that we're letting perfection be the enemy of good. This PR would change basically nothing for those that already use the Dracut default, and would be a material improvement for those that don't.

Bikeshedding on The One True Initramfs Name accomplishes nothing. This isn't even the right forum for that discussion! I can promise you that Debian maintainers don't give a seconds thought to a random GitHub discussion.

**Comment by Conan-Kudo** [comment]
> @Conan-Kudo How do you feel about the idea of opening an issue in the https://github.com/uapi-group/specifications/issues repo and try to sort the naming out there ?
> 

I could try, I'm not sure the UAPI group cares about this kind of detail, but it's worth a shot.

> > Debian family, which I would like to see if they could be convinced to use the initramfs-$KVER.img form too.
> 
> This would have to include changing initramfs-tools as well I am afraid.
> 
> I think e.g. grub understand wildcards, so maybe distro's default bootloader should be just using the following to help with the transition ? `initrd /boot/initr*-$KVER*`

I'm not sure that grub itself cares? This is usually driven by whatever hook is used to generate the boot entry.

**Comment by Conan-Kudo** [comment]
> It's unfortunate that we're letting perfection be the enemy of good. This PR would change basically nothing for those that already use the Dracut default, and would be a material improvement for those that don't.
> 
> Bikeshedding on The One True Initramfs Name accomplishes nothing. This isn't even the right forum for that discussion! I can promise you that Debian maintainers don't give a seconds thought to a random GitHub discussion.

This isn't a random discussion, and the maintainer of Dracut in Debian is actually present in here too (@Mrfai) as is the Ubuntu maintainer (@bdrung).

And there is something to accomplish here: making life easier for people building _on top_ of distributions, particularly when making their own images and making their own tools for doing so.

**Comment by Conan-Kudo** [comment]
(Also, as an aside, kiwi is not a Fedora project, it came from openSUSE. Again, I reiterate, I am an upstream developer in multiple concerns across multiple Linux distributions, but I am _not_ involved in Fedora at that part of the stack.)

**Comment by zdykstra** [comment]
> > It's unfortunate that we're letting perfection be the enemy of good. This PR would change basically nothing for those that already use the Dracut default, and would be a material improvement for those that don't.
> > 
> > Bikeshedding on The One True Initramfs Name accomplishes nothing. This isn't even the right forum for that discussion! I can promise you that Debian maintainers don't give a seconds thought to a random GitHub discussion.
> 
> This isn't a random discussion, and the maintainer of Dracut in Debian is actually present in here too (@Mrfai) as is the Ubuntu maintainer (@bdrung).
> 
> And there is something to accomplish here: making life easier for people building _on top_ of distributions, particularly when making their own images and making their own tools for doing so.

That those maintainers are here is largely immaterial. The scope of this PR is quite small and it solves an immediate need with no appreciable maintenance burden.

Wouldn't it be more appropriate to brain storm the next step - getting a common name format between distributions - in a better forum, and not a PR that solves an immediate problem?

**Comment by Conan-Kudo** [comment]
@LaszloGombos I've filed a ticket with the UAPI group folks: https://github.com/uapi-group/specifications/issues/100

**Comment by Mrfai** [comment]
In Debian the default initrd tool is not dracut but
initramfs-tools. The later is used much more often that dracut, nearly
everybody is using initramfs-tools and we do not know when Debian will
switch to dracut as default. It will help if Ubuntu also likes to do
the switch.

I like to have the initrd name in sync with the name used by
initramfs-tools and discuss with them when to switch to a new naming
scheme.

But I'm fine with any solution, either applying a patch for the Debian
package or setting a variable in a config file.

So feel free to decide what you think is better.
-- 
regards Thomas

**Comment by schaefi** [comment]
> That those maintainers are here is largely immaterial. The scope of this PR is quite small and it solves an immediate need with no appreciable maintenance burden.

I would agree to that. The proposed change looks generally useful and is imho not in conflict with the initiative to classify a common name that every distribution should use. From a neutral tooling position I always enjoy tools that give users enough
flexibility to solve a use case without the need of extra polishing code prior or after the tool call. In this case having the opportunity to influence the output file name is valuable.

> Wouldn't it be more appropriate to brain storm the next step - getting a common name format between distributions - in a better forum, and not a PR that solves an immediate problem?

I'm fully behind you with that. Not sure if "appropriate" matches best in this thread as I think both topics, the common name, as well as the proposed patch here are both important to be discussed. I see the patch here as a good value for the tool itself and a door opener for the conversation about a standardized name. In the long run I agree with you that a standard name defined and accepted is more valuable as it would simplify the use of the tool and the implementation.

I myself come from the appliance building area and always wanted to provide an image builder that works for all distributions. You can imagine one of the biggest hurdles were the "unneeded" differences between the distributions which could be solved by accepted standards and consistency. Linux has improved a lot here and dracut has provided a great part to a possible standard for producing the initrd. I remember times when every distribution had their own coding for "mkinitrd". Today most of them are using dracut... but not all e.g Arch

So what I'd like to say, even if we have an accepted common name and written up in UAPI there is still the individual choice of the distributions to respect it or not. I think this cannot be enforced and usually causes trouble if you try :)

Thus a +1 for the proposed patch from my side.

Thanks

**Comment by bdrung** [comment]
After this discussion here and my initial +1 on this change, I changed my mind: The proper fix for the issue is to change the initramfs name in Debian/Ubuntu and drop the rename patch. Merging this change would just increase the complexity without a benefit. I see no reason why different distributions have different names for their initramfs file. It looks like the names currently in place is just it developed in the past. Having `initramfs-${KERNEL_VERSION}.img` is fine for me.

**Comment by bdrung** [comment]
> This is an initramfs-tools question (sorry!), but I am wondering @bdrung if we can interpret your opinion such that you're planning to help with changes in initramfs-tools as well to go through this transition ?

Yes. This kind of change could land in Ubuntu 24.10.

**Pull request closed without merging**